pkg/sshd/server.go

```diff
@@ -232,8 +237,29 @@ func (s *server) answer(channel ssh.Channel, requests <-chan *ssh.Request, sshCo
 				}
 
 				repoName := parts[1]
-				s.cleanerRef.Lock()
-				if err := s.pushLock.Lock(repoName, time.Duration(0)); err != nil {
+				errConcurrentPush := errors.New("concurrent push")
+				err := s.withCleanerLock(func() error {
+					if err := s.pushLock.Lock(repoName, time.Duration(0)); err != nil {
+						return errConcurrentPush
+					}
+
+					req.Reply(true, nil) // We processed. Yay.
+
+					cxt.Put("channel", channel)
+					cxt.Put("request", req)
+					cxt.Put("operation", parts[0])
+					cxt.Put("repository", parts[1])
+					sshGitReceive := cxt.Get("route.sshd.sshGitReceive", "sshGitReceive").(string)
+					err := router.HandleRequest(sshGitReceive, cxt, true)
+					if err := s.pushLock.Unlock(repoName, time.Duration(0)); err != nil {
+						log.Errf(s.c, "unable to unlock repository lock for %s (%s)", repoName, err)
+						// TODO: this is an important error case that needs to be covered
```

Comment:
do we have an issue for this? This is definitely a big issue, but it's better than v1's file lock implementation, at least.